src/engine/SpatialJoinAlgorithms.cpp

```diff
@@ -255,11 +286,126 @@ Result SpatialJoinAlgorithms::BaselineAlgorithm() {
+    cfg.sepCrosses = std::string{static_cast<char>(SpatialJoinType::CROSSES)};
+    cfg.pairEnd = "";
+    cfg.useBoxIds = true;
+    cfg.useArea = true;
+    cfg.useOBB = false;
+    cfg.useCutouts = true;
+    cfg.useDiagBox = true;
+    cfg.useFastSweepSkip = true;
+    cfg.useInnerOuter = false;
+    cfg.noGeometryChecks = false;
+    cfg.withinDist = withinDist;
+    cfg.writeRelCb = [&results, &resultDists, joinTypeVal](
+                         size_t t, const char* a, const char* b,
+                         const char* pred) {
+      if (joinTypeVal == SpatialJoinType::WITHIN_DIST) {
+        results[t].push_back({std::atoi(a), std::atoi(b)});
+        resultDists[t].push_back(atof(pred));
+      } else if (pred[0] == static_cast<char>(joinTypeVal)) {
+        results[t].push_back({std::atoi(a), std::atoi(b)});
+      }
+    };
+    cfg.logCb = {};
+    cfg.statsCb = {};
+    cfg.sweepProgressCb = {};
+    return cfg;
+  }();
+
+  sj::Sweeper sweeper(sweeperCfg, ".", "", "spatialjoin");
+
+  ad_utility::Timer tParse{ad_utility::Timer::Started};
+
+  // Parse the geometries from the left and right input table, starting with the
+  // smaller one. Compute the bounding box of the smaller table (appropriately
+  // inflated for `WITHIN_DIST` joins) and only add those geometries from the
+  // larger table that intersect this bounding box.
+  if (idTableLeft->size() < idTableRight->size()) {
+    auto box = libspatialjoinParse(false, idTableLeft, leftJoinCol, sweeper,
+                                   NUM_THREADS);
+    sweeper.setFilterBox(box);
+    libspatialjoinParse(true, idTableRight, rightJoinCol, sweeper, NUM_THREADS);
+  } else {
+    auto box = libspatialjoinParse(true, idTableRight, rightJoinCol, sweeper,
+                                   NUM_THREADS);
+    sweeper.setFilterBox(box);
+    libspatialjoinParse(false, idTableLeft, leftJoinCol, sweeper, NUM_THREADS);
+  }
+
+  // Flush the geometries and add the time for parsing and processing the
+  // geometries to the runtime information.
+  sweeper.flush();
+  spatialJoin_.value()->runtimeInfo().addDetail("time for reading geometries",
+                                                tParse.msecs().count());
+
+  // Now do the sweep, which performs the actual spatial join.
+  ad_utility::Timer tSweep{ad_utility::Timer::Started};
+  sweeper.sweep();
+  spatialJoin_.value()->runtimeInfo().addDetail("time for spatialjoin sweep",
+                                                tSweep.msecs().count());
+  ad_utility::Timer tCollect{ad_utility::Timer::Started};
+
+  // Collect the results and add them to the result table. For `WITHIN_DIST`,
+  // also add the distance for each pair of objects in the result.
+  for (size_t t = 0; t < NUM_THREADS; t++) {
+    for (size_t i = 0; i < results[t].size(); i++) {
+      const auto& res = results[t][i];
+      double dist = 0;
+      if (withinDist >= 0) dist = resultDists[t][i];
```

Comment:
Do you really need the hack with the non-negative `withinDist` here (see also the corresponding `withinDist = -1`  above). Why not add the distance if and only if the join type is `WITHIN_DIST`?

Reply:
I replaced the `= -1` check with a check for `WITHIN_DIST` here, but `maxDist` must be set to -1 if `WITHIN_DIST` is not used (`libspatialjoin` will compute `WITHIN_DIST` if `maxDist >= 0`)